### PR TITLE
#4231: improve unary add, sub, mul and div implementation in SFPU. Add complex polar operator

### PR DESCRIPTION
**Pull request by muthutt**
#0: remove cruft and fine-tune the unary ops to improve performance and cut down kernel launches

#0: move add, sub, div, mul binops with scalar tied in as full featured unary op on sfpu
    remove RDIV from the unary kernel since it has poor accuracy
#0: remove rdiv from the eltwise_binop_unary_with_param code for GS like in WH

#4231: add polar hooks and python binding
     : polar operator for mistral
     : add doc for polar
     : polar test

#0: update soft sign implementation to be faster

#0: softsign test added and some profiling result for refactoring
image

### CODE REVIEW

File: tt_metal/hw/ckernels/grayskull/common/inc/ckernel_sfpu_binop_with_unary.h
Comment (tt-aho) [inline tt_metal/hw/ckernels/grayskull/common/inc/ckernel_sfpu_binop_with_unary.h:60]: 4 space indentation per level

File: tt_metal/hw/ckernels/grayskull/common/inc/ckernel_sfpu_binop_with_unary.h
Comment (tt-aho) [inline tt_metal/hw/ckernels/grayskull/common/inc/ckernel_sfpu_binop_with_unary.h:86]: 4 space indentation

File: tt_metal/hw/ckernels/grayskull/common/inc/ckernel_sfpu_binop_with_unary.h
Comment (tt-aho) [inline tt_metal/hw/ckernels/grayskull/common/inc/ckernel_sfpu_binop_with_unary.h:35]: Host should pass in inv_parameter value instead of computing it in kernel runtime for better perf

File: tt_metal/hw/ckernels/grayskull/llk_lib/llk_math_eltwise_unary_sfpu_binop_with_scalar.h
Comment (tt-aho) [inline tt_metal/hw/ckernels/grayskull/llk_lib/llk_math_eltwise_unary_sfpu_binop_with_scalar.h:19]: 4 space indentation

File: tt_metal/hw/ckernels/grayskull/common/inc/ckernel_sfpu_binop_with_unary.h
Comment (tt-aho) [inline tt_metal/hw/ckernels/grayskull/common/inc/ckernel_sfpu_binop_with_unary.h:94]: This is not needed? init should be empty?

File: tt_metal/hw/ckernels/wormhole_b0/common/inc/ckernel_sfpu_binop_with_unary.h
Comment (tt-aho) [inline tt_metal/hw/ckernels/wormhole_b0/common/inc/ckernel_sfpu_binop_with_unary.h:34]: Same comments as gs file for indentation / issues

File: tt_metal/hw/ckernels/wormhole_b0/llk_lib/llk_math_eltwise_unary_sfpu_binop_with_scalar.h
Comment (tt-aho) [inline tt_metal/hw/ckernels/wormhole_b0/llk_lib/llk_math_eltwise_unary_sfpu_binop_with_scalar.h:19]: Indentation

File: tt_metal/include/compute_kernel_api/eltwise_unary/binop_with_scalar.h
Comment (tt-aho) [inline tt_metal/include/compute_kernel_api/eltwise_unary/binop_with_scalar.h:29]: Wrong description (not identity)

File: tt_metal/include/compute_kernel_api/eltwise_unary/binop_with_scalar.h
Comment (tt-aho) [inline tt_metal/include/compute_kernel_api/eltwise_unary/binop_with_scalar.h:58]: Indentation

File: tt_eager/tt_dnn/op_library/eltwise_unary/eltwise_unary_op.cpp
Comment (tt-aho) [inline tt_eager/tt_dnn/op_library/eltwise_unary/eltwise_unary_op.cpp:129]: indentation

File: tt_eager/tt_dnn/op_library/composite/composite_ops.cpp
Comment (tt-aho) [inline tt_eager/tt_dnn/op_library/composite/composite_ops.cpp:382]: Indentation

File: tt_eager/tt_dnn/op_library/complex/complex_ops.cpp
Comment (tt-aho) [inline tt_eager/tt_dnn/op_library/complex/complex_ops.cpp:181]: Indentation

**Review by tt-aho — CHANGES_REQUESTED** [review]
Do you have profiling data of the original bcast implementation vs the new sfpu implementation for comparison?

**Review by eyonland — APPROVED** [review]

**Review by mywoodstock — APPROVED** [review]

**Review by tt-aho — CHANGES_REQUESTED** [review]
Please reformat the following three files you used emacs to format:
```
tt_metal/hw/ckernels/grayskull/common/inc/ckernel_sfpu_binop_with_unary.h
tt_metal/hw/ckernels/wormhole_b0/common/inc/ckernel_sfpu_binop_with_unary.h
tt_metal/include/compute_kernel_api/eltwise_unary/binop_with_scalar.h
```
The emacs formatting is indenting namespaces and doesn't really follow our guidelines for formatting. I would suggest using clang-format if you want to use a formatter as we have a config file in the repo that follows the formatting guidelines.

Also, not sure if I misunderstood/missed it but the graph you added to the OP does not have profiling data for the modified ops? It is just the other sfpu ops which shouldn't be affected by this PR?
Do you have numbers for the bcast vs sfpu implementations you could share?

File: tests/tt_eager/python_api_testing/unit_testing/test_complex.py
Comment (tt-aho) [reply tests/tt_eager/python_api_testing/unit_testing/test_complex.py]: My comment was unrelated to the current implementation issues or accuracy.

It was just based on what seems like unnecessarily creating the input tensors for both pt and tt in a complex wrapper when they are/should just be regular tensors. For example the line `tt_cpu = torch.polar(x.real, x.imag)` the inputs should be the abs and angle, and returns a complex tensor with real and imag values, and I find it confusing that you are basically passing a complex tensor as input when it's not supposed to be a real complex tensor.

This is overall minor since it is not related to implementation so not blocking the pr on this.

File: tt_eager/tt_dnn/op_library/complex/complex_ops.cpp
Comment (tt-aho) [reply tt_eager/tt_dnn/op_library/complex/complex_ops.cpp]: Are you planning on rearranging the order of operations as suggested?

File: tests/tt_eager/python_api_testing/unit_testing/test_complex.py
Comment (tt-aho) [inline tests/tt_eager/python_api_testing/unit_testing/test_complex.py:831]: You need to assert if it passed here based on the return value.

```
    real_passing, real_output = comp_allclose(tt_cpu_real, tt_dev_real, 0.1, 1e-3)
    logger.info(real_output)
    imag_passing, imag_output = comp_allclose(tt_cpu_imag, tt_dev_imag, 0.1, 1e-3)
    logger.info(imag_output)
    assert real_passing and imag_passing
```

**Review by tt-aho — APPROVED** [review]
lgtm. Would be good to get profiling info when you can to compare which version is better as a standalone op but having the sfpu implementation is useful either way for fusing.

I am approving now but you need to add the asserts to the test and run before merge

**Pull request merged**
